### PR TITLE
Add `Launch at Login` setting for macOS and Windows

### LINKED ISSUE #773: Add "Launch at Login" menu item
https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows
https://electronjs.org/docs/api/app#appgetloginitemsettingsoptions-macos-windows

### PR DESCRIPTION
**Pull request by whitecrownclown**
@sindresorhus 
Fixes #773 

Hello!
I've added the following : 

- Launch at login menu item
- Launch as hidden menu item (disabled if launch at login is false)
- Tested on macOS 10.14.3

Feedback is welcome, thanks!

### CODE REVIEW

**Comment by sindresorhus** [comment]
You shouldn’t store the states in the config store. Just fetch the actual states from the `getLogin...` method when needed. That way the settings don’t get out of sync.

File: source/menu.ts [hunk 1]
```diff
@@ -13,6 +13,23 @@ import config from './config';
 import {sendAction, showRestartDialog} from './util';
 import {generateSubmenu as generateEmojiSubmenu} from './emoji';
 
+const appFolder = path.dirname(process.execPath);
+const updateExe = path.resolve(appFolder, '..', 'Update.exe');
```
Comment (sindresorhus) [inline source/menu.ts]: ?

File: source/menu.ts [hunk 2]
```diff
@@ -190,6 +209,24 @@ Press Command/Ctrl+R in Caprine to see your changes.
 				focusedWindow.setAlwaysOnTop(item.checked);
 			}
 		},
+		{
+			type: 'checkbox',
+			label: 'Launch at login',
```
Comment (sindresorhus) [inline source/menu.ts]: `label` should be first

File: source/menu.ts (hunk 1 above)
Comment (whitecrownclown) [reply source/menu.ts]: I only moved those outside of the function definitions since they're not updating at all

Comment (sindresorhus) [reply source/menu.ts]: No, I mean. Can you explain why this is needed?

**Comment by sindresorhus** [comment]
> * Launch as hidden menu item (disabled if launch at login is false)

I don't see the point of having a setting for this. We should just launch hidden.

File: source/menu.ts [hunk 3]
```diff
@@ -13,6 +13,23 @@ import config from './config';
 import {sendAction, showRestartDialog} from './util';
 import {generateSubmenu as generateEmojiSubmenu} from './emoji';
 
+const appFolder = path.dirname(process.execPath);
+const updateExe = path.resolve(appFolder, '..', 'Update.exe');
+
+function setLoginSettings({openAtLogin, openAsHidden}: AppLoginSettings): void {
```
Comment (sindresorhus) [inline source/menu.ts]: Just use the type for `setLoginItemSettings` instead of defining your own.

Comment (whitecrownclown) [reply source/menu.ts]: Ok, i did look for it at first, didn't find it, i'll have a closer look.

File: source/menu.ts (hunk 1 above)
Comment (whitecrownclown) [reply source/menu.ts]: Initially it was for : [To work with Electron's autoUpdater on Windows](https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows).

Comment (whitecrownclown) [reply source/menu.ts]: I removed it for now, as i cannot really test it and also don't have access to windows.

**Comment by whitecrownclown** [comment]
@sindresorhus I've updated the PR again.

* Added the ts definition from electron
* Removed `Launch as Hidden`; defaults to the same value as `Launch at Login`

File: source/menu.ts [hunk 4]
```diff
@@ -13,6 +13,14 @@ import config from './config';
 import {sendAction, showRestartDialog} from './util';
 import {generateSubmenu as generateEmojiSubmenu} from './emoji';
 
+function setLoginSettings(settings: LoginItemSettings): void {
+	app.setLoginItemSettings(settings);
+}
+
+function getLoginSettings(): LoginItemSettings {
+	return app.getLoginItemSettings();
+}
```
Comment (sindresorhus) [inline source/menu.ts]: These wrappers are now moot.

File: source/menu.ts [hunk 5]
```diff
@@ -190,6 +200,14 @@ Press Command/Ctrl+R in Caprine to see your changes.
 				focusedWindow.setAlwaysOnTop(item.checked);
 			}
 		},
+		{
+			label: 'Launch at Login',
+			type: 'checkbox',
+			checked: settings.openAtLogin,
+			click(item) {
+				setLoginSettings({...settings, openAtLogin: item.checked, openAsHidden: item.checked});
```
Comment (sindresorhus) [inline source/menu.ts]: I don't think you need `...settings`. And `openAsHidden: item.checked` needs to be updated.

Comment (whitecrownclown) [reply source/menu.ts]: I thought so as well, but i'm getting : `Type '{ openAtLogin: boolean; openAsHidden: boolean; }' is missing the following properties from type 'LoginItemSettings': wasOpenedAtLogin, wasOpenedAsHidden, restoreState`

File: source/menu.ts (hunk 4 above)
Comment (whitecrownclown) [reply source/menu.ts]: Yes, will update

File: source/menu.ts (hunk 5 above)
Comment (whitecrownclown) [reply source/menu.ts]: Disregard, it's no longer applicable since i've removed the wrappers.

**Comment by whitecrownclown** [comment]
@sindresorhus Updated again, however i'm unsure what you mean by this:

>  And openAsHidden: item.checked needs to be updated.

Do you mean to set it to `true` always? It is being updated with the same value as `Launch at Login`. I've noticed if `Launch at Login` is `false` it will not set `Launch as Hidden` as `true`, but `false` as well.

**Comment by sindresorhus** [comment]
> Do you mean to set it to `true` always?

Yes, setting it to `true` when `openAtLogin: false` should have no effect, so might as well make it simpler.

File: source/menu.ts [hunk 6]
```diff
@@ -190,6 +190,14 @@ Press Command/Ctrl+R in Caprine to see your changes.
 				focusedWindow.setAlwaysOnTop(item.checked);
 			}
 		},
+		{
+			label: 'Launch at Login',
```
Comment (sindresorhus) [inline source/menu.ts]: This feature only works on macOS and Windows, so you have to limit it to those system with the `visible` option.

https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows

Comment (whitecrownclown) [reply source/menu.ts]: Yep, I missed that, thanks!

**Comment by sindresorhus** [comment]
Thanks :)

**Comment by blargle-ragequit** [comment]
Why is this not on Linux?

**Comment by CvX** [comment]
@blargle-ragequit Because the Electron API doesn’t support it: https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows

Here’s an open issue at the Electron repo: https://github.com/electron/electron/issues/15198

**Pull request merged**